File: pkg/models/collections.go
package models

import "time"

// CollectionRole captures membership roles within a collection.
type CollectionRole string

const (
	// CollectionRoleOwner can edit, invite, remove members, delete the collection.
	CollectionRoleOwner CollectionRole = "owner"
	// CollectionRoleMember can read items and run queries they have source access to.
	CollectionRoleMember CollectionRole = "member"
)

// Collection groups saved queries across teams. Each user has a single
// auto-created personal collection (is_personal = true); other collections
// are shared with explicit member invites.
type Collection struct {
	ID          int            `json:"id"`
	Name        string         `json:"name"`
	Description string         `json:"description,omitempty"`
	IsPersonal  bool           `json:"is_personal"`
	CreatedBy   *UserID        `json:"created_by,omitempty"`
	CallerRole  CollectionRole `json:"caller_role"`
	MemberCount int            `json:"member_count"`
	ItemCount   int            `json:"item_count"`
	CreatedAt   time.Time      `json:"created_at"`
	UpdatedAt   time.Time      `json:"updated_at"`
}

// CollectionMember is one membership row.
type CollectionMember struct {
	CollectionID int            `json:"collection_id"`
	UserID       UserID         `json:"user_id"`
	Role         CollectionRole `json:"role"`
	AddedBy      *UserID        `json:"added_by,omitempty"`
	CreatedAt    time.Time      `json:"created_at"`
	Email        string         `json:"email,omitempty"`
	FullName     string         `json:"full_name,omitempty"`
}

// CollectionItem is a saved query included in a collection. Runnable indicates
// whether the requesting user has source access to actually run it; non-runnable
// items still surface in listings (with a lock icon on the UI side) so members
// can see the curated list even if some queries are out of reach.
type CollectionItem struct {
	CollectionID int        `json:"collection_id"`
	SortOrder    int        `json:"sort_order"`
	AddedBy      *UserID    `json:"added_by,omitempty"`
	ItemAddedAt  time.Time  `json:"item_added_at"`
	Query        SavedQuery `json:"query"`
	Runnable     bool       `json:"runnable"`
}

// CreateCollectionRequest is the JSON body for POST /api/v1/collections.
type CreateCollectionRequest struct {
	Name        string `json:"name" validate:"required"`
	Description string `json:"description"`
}

// UpdateCollectionRequest is the JSON body for PUT /api/v1/collections/:id.
type UpdateCollectionRequest struct {
	Name        string `json:"name"`
	Description string `json:"description"`
}

// AddCollectionMemberRequest is the JSON body for POST /api/v1/collections/:id/members.
type AddCollectionMemberRequest struct {
	UserID UserID         `json:"user_id" validate:"required"`
	Role   CollectionRole `json:"role" validate:"required"`
}

// AddCollectionItemRequest is the JSON body for POST /api/v1/collections/:id/items.
type AddCollectionItemRequest struct {
	SavedQueryID int `json:"saved_query_id" validate:"required"`
	SortOrder    int `json:"sort_order"`
}

File: frontend/src/api/collections.ts
import { apiClient } from "./apiUtils";
import type { SavedQuery } from "./savedQueries";

export type CollectionRole = "owner" | "editor" | "member";

export interface Collection {
  id: number;
  name: string;
  description?: string;
  is_personal: boolean;
  created_by: number;
  caller_role?: CollectionRole;
  member_count: number;
  item_count: number;
  created_at: string;
  updated_at: string;
}

export interface CollectionMember {
  collection_id: number;
  user_id: number;
  role: CollectionRole;
  added_by?: number | null;
  created_at: string;
  email?: string;
  full_name?: string;
}

export interface CollectionItem {
  collection_id: number;
  sort_order: number;
  added_by?: number | null;
  item_added_at: string;
  query: SavedQuery;
  runnable: boolean;
}

export interface CreateCollectionRequest {
  name: string;
  description?: string;
}

export interface UpdateCollectionRequest {
  name: string;
  description?: string;
}

export interface AddCollectionMemberRequest {
  user_id: number;
  role: CollectionRole;
}

export interface AddCollectionItemRequest {
  saved_query_id: number;
  sort_order?: number;
}

export const collectionsApi = {
  list: () => apiClient.get<Collection[]>("/collections"),
  get: (id: number) => apiClient.get<Collection>(`/collections/${id}`),
  create: (payload: CreateCollectionRequest) =>
    apiClient.post<Collection>("/collections", payload),
  update: (id: number, payload: UpdateCollectionRequest) =>
    apiClient.put<Collection>(`/collections/${id}`, payload),
  delete: (id: number) =>
    apiClient.delete<{ message: string }>(`/collections/${id}`),

  listMembers: (id: number) =>
    apiClient.get<CollectionMember[]>(`/collections/${id}/members`),
  addMember: (id: number, payload: AddCollectionMemberRequest) =>
    apiClient.post<{ message: string }>(`/collections/${id}/members`, payload),
  removeMember: (id: number, userId: number) =>
    apiClient.delete<{ message: string }>(`/collections/${id}/members/${userId}`),

  listItems: (id: number) =>
    apiClient.get<CollectionItem[]>(`/collections/${id}/items`),
  addItem: (id: number, payload: AddCollectionItemRequest) =>
    apiClient.post<{ message: string }>(`/collections/${id}/items`, payload),
  removeItem: (id: number, queryId: number) =>
    apiClient.delete<{ message: string }>(`/collections/${id}/items/${queryId}`),
};
